### PR TITLE
feat(tracer): stats to OTLP histogram conversion for OTLP span metrics

### PR DESCRIPTION
**Pull request by rachelyangdog**
### OTLP Span Metrics — Part 2 of 4: Stats → OTLP Conversion

Implements the conversion layer that turns the tracer's internal `ClientStatsPayload` into an OTLP `ExportMetricsServiceRequest`.

**Stack order (merge in sequence):**
1. `otlp-span-metrics-config` — configuration (PR #4895)
2. **This PR** — ClientStatsPayload → OTLP protobuf conversion
3. `otlp-span-metrics-transport` — OTLP HTTP exporter
4. `otlp-span-metrics-wireup` — tracer integration and system-test fixes

#### What this PR does

Adds `ddtrace/tracer/stats_to_otlp_metrics.go` which converts a `ClientStatsPayload` bucket into an OTLP histogram metric named `traces.span.sdk.metrics.duration`:

- Service identity (`service.name`, `deployment.environment.name`, `service.version`) goes on the OTLP Resource
- Per-span dimensions (`datadog.operation.name`, `span.kind`, `http.request.method`, `http.route`, `rpc.method`, etc.) go on each data point
- Duration values are converted from nanoseconds to seconds and bucketed using the OTel Span Metrics Connector default boundaries
- Datadog peer-tag keys are mapped to OTel semantic-convention names (e.g. `grpc.method.name` → `rpc.method`)
- OTel-semantics mode (`DD_TRACE_OTEL_SEMANTICS_ENABLED`) substitutes OTel attribute names for Datadog ones where they differ

#### Testing
- Unit tests in `stats_to_otlp_metrics_test.go` covering attribute mapping, bucket boundaries, OTel-semantics mode, and edge cases

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] New code is free of linting errors. You can check this by running `make lint` locally.
- [ ] New code doesn't break existing tests. You can check this by running `make test` locally.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] All generated files are up to date. You can check this by running `make generate` locally.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild. Make sure all nested modules are up to date by running `make fix-modules` locally.

Unsure? Have a question? Request a review!

### CODE REVIEW

**Comment by rachelyangdog** [comment]
@codex review

**Comment by rachelyangdog** [comment]
@codex review

File: ddtrace/tracer/stats_to_otlp_metrics.go [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+			Resource:     resource,
+			ScopeMetrics: scopeMetrics,
+		},
+	}
+}
+
+// buildMetricsResource constructs the OTLP Resource for the metrics payload.
+// Includes SDK identification, service identity, optional host.name, and datadog.* process-tag
+// attributes (default mode only).
+func buildMetricsResource(cfg *internalconfig.Config, payload *pb.ClientStatsPayload, otelMode bool) *otlpresource.Resource {
```
Comment (mtoffl01) [inline ddtrace/tracer/stats_to_otlp_metrics.go]:
```suggestion
func buildMetricsResource(payload *pb.ClientStatsPayload, otelMode bool, reportHostname bool, hostname string) *otlpresource.Resource {
```

This way it matches buildGroupDataPoints, and then both functions take in a snapshot of `cfg` values rather than `cfg` itself.

File: ddtrace/tracer/stats_to_otlp_metrics.go [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+// spanMetricBounds are the fixed histogram bucket boundaries in seconds (16 boundaries, 17 buckets).
+// These match the OTel Span Metrics Connector defaults, scaled from milliseconds to seconds.
+var spanMetricBounds = []float64{0.002, 0.004, 0.006, 0.008, 0.01, 0.05, 0.1, 0.2, 0.4, 0.8, 1, 1.4, 2, 5, 10, 15}
+
+// BuildOTLPMetricsRequest converts a ClientStatsPayload into a slice of OTLP ResourceMetrics.
+// Service identity (service.name, service.version, deployment.environment.name) is placed on the
+// Resource. When a grouped-stats entry carries a different service than the payload's configured
+// default, service.name is additionally emitted as a data-point attribute. The resulting histogram
+// uses DELTA temporality. Returns nil when the payload produces no data points.
+func BuildOTLPMetricsRequest(payload *pb.ClientStatsPayload, cfg *internalconfig.Config) []*otlpmetrics.ResourceMetrics {
```
Comment (mtoffl01) [inline ddtrace/tracer/stats_to_otlp_metrics.go]: [Unless a subsequent PR calls this from outside the `tracer` package] this can be unexported so as not to pollute the public API surface: `buildOTLPMetricsRequest`

File: ddtrace/tracer/stats_to_otlp_metrics.go [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+	internalconfig "github.com/DataDog/dd-trace-go/v2/internal/config"
+	"github.com/DataDog/dd-trace-go/v2/internal/log"
+	"github.com/DataDog/dd-trace-go/v2/internal/version"
+)
+
+const spanDurationMetricName = "traces.span.sdk.metrics.duration"
+
+// spanMetricBounds are the fixed histogram bucket boundaries in seconds (16 boundaries, 17 buckets).
+// These match the OTel Span Metrics Connector defaults, scaled from milliseconds to seconds.
+var spanMetricBounds = []float64{0.002, 0.004, 0.006, 0.008, 0.01, 0.05, 0.1, 0.2, 0.4, 0.8, 1, 1.4, 2, 5, 10, 15}
```
Comment (mtoffl01) [inline ddtrace/tracer/stats_to_otlp_metrics.go]: nit: you can define this as an array with a fixed length instead -- assuming it will never grow.

**Review by mtoffl01 — APPROVED** [review]
some nits to address, then LGTM.

File: ddtrace/tracer/stats_to_otlp_metrics.go (hunk 2 above)
Comment (rachelyangdog) [reply ddtrace/tracer/stats_to_otlp_metrics.go]: @datadog fix this

File: ddtrace/tracer/stats_to_otlp_metrics.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+	}
+	if !otelMode {
+		if payload.RuntimeID != "" {
+			attrs = append(attrs, otlpKeyValue("datadog.runtime_id", otlpStringValue(payload.RuntimeID)))
+		}
+		if payload.ProcessTags != "" {
+			for tag := range strings.SplitSeq(payload.ProcessTags, ",") {
+				parts := strings.SplitN(tag, ":", 2)
+				if len(parts) == 2 && parts[0] != "" {
+					attrs = append(attrs, otlpKeyValue("datadog."+parts[0], otlpStringValue(parts[1])))
```
Comment (darccio) [inline ddtrace/tracer/stats_to_otlp_metrics.go:116]: This loop blindly prefixes any tag key with `datadog.`. If `ProcessTags` ever contained `runtime_id:...`, you'd get a second `datadog.runtime_id` alongside the explicit one at line 99, causing duplicate resource keys (OTLP says keys "SHOULD" be unique). Also nothing constrains the key set, so a tag like `service:foo` becomes `datadog.service`. In practice process tags are `entrypoint.*` today, so this is latent.

File: ddtrace/tracer/stats_to_otlp_metrics.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+		attrs = append(attrs, otlpKeyValue("service.version", otlpStringValue(payload.Version)))
+	}
+	if payload.Env != "" {
+		attrs = append(attrs, otlpKeyValue("deployment.environment.name", otlpStringValue(payload.Env)))
+	}
+	if reportHostname {
+		if hostname != "" {
+			attrs = append(attrs, otlpKeyValue("host.name", otlpStringValue(hostname)))
+		}
+	}
```
Comment (darccio) [inline ddtrace/tracer/stats_to_otlp_metrics.go]: This nested conditional can be collapsed to a single `&&`:


```suggestion
	if reportHostname && hostname != "" {
		attrs = append(attrs, otlpKeyValue("host.name", otlpStringValue(hostname)))
	}
```

File: ddtrace/tracer/stats_to_otlp_metrics.go [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+	dp := &otlpmetrics.HistogramDataPoint{
+		StartTimeUnixNano: startNs,
+		TimeUnixNano:      endNs,
+		Count:             count,
+		Sum:               &sum,
+		Min:               &minSec,
+		Max:               &maxSec,
+		ExplicitBounds:    spanMetricBounds[:],
+		BucketCounts:      bucketCounts,
```
Comment (darccio) [inline ddtrace/tracer/stats_to_otlp_metrics.go:160]: If I understood this correctly, `Count` and `BucketCounts` seem to come from two different sources. `dp.Count = exactCount` (the exact `Hits-Errors` / `Errors` counter), but `dp.BucketCounts` is the sketch's histogram, which sums to `sketchCount`. The OTLP spec requires `sum of bucket_counts == count` ([source](https://github.com/open-telemetry/opentelemetry-proto-go/blob/otlp/v1.10.0/otlp/metrics/v1/metrics.pb.go#L1189)). In the normal path these should match (the OK sketch seems to have exactly `Hits-Errors` values, and `LogCollapsingLowestDense` preserves total count), so it's probably fine in the common case but the code never reconciles or asserts it, and the two values are computed independently. If they ever diverge (sampling, partial sketches, rounding drift), this will emit a histogram that violates the invariant.

File: ddtrace/tracer/stats_to_otlp_metrics.go [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+	return dp
+}
+
+// buildDataPointAttributes returns OTLP data-point attributes; adds service.name for non-default services.
+func buildDataPointAttributes(gs *pb.ClientGroupedStats, isError bool, defaultService string, otelMode bool) []*otlpcommon.KeyValue {
+	var attrs []*otlpcommon.KeyValue
+
+	// OTel semantic-convention attributes.
+	if gs.Resource != "" {
+		attrs = append(attrs, otlpKeyValue("span.name", otlpStringValue(gs.Resource)))
```
Comment (darccio) [inline ddtrace/tracer/stats_to_otlp_metrics.go:176]: Confirm the `span.name` ← `Resource` mapping, and that dropping the operation name in OTel mode is intended.

Two coupled decisions here:
- `span.name` is populated from `gs.Resource` (e.g. `GET /users/{id}`), not `gs.Name`.
- `gs.Name` (the Datadog operation name, e.g. `web.request`) is only emitted as `datadog.operation.name`, which is suppressed in OTel mode (lines 209-210).

Net effect: in OTel-semantics mode, the operation name is **not present on the data point at all** — no `span.name`, no `datadog.operation.name`, nothing carries it. Only the resource string survives.

Is that the intended behavior per SEMCON-1093? If the RFC does specify `span.name = Resource`, a one-line comment pointing at it would save the next reader the same double-take. If the operation name is meant to be retrievable in OTel mode, it needs its own OTel-semconv attribute.

File: ddtrace/tracer/stats_to_otlp_metrics.go [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+			attrs = append(attrs, otlpKeyValue("rpc.response.status_code", otlpIntValue(code)))
+		} else {
+			attrs = append(attrs, otlpKeyValue("rpc.response.status_code", otlpStringValue(gs.GRPCStatusCode)))
+		}
+	}
+	statusCode := "Ok"
+	if isError {
+		statusCode = "Error"
+	}
+	attrs = append(attrs, otlpKeyValue("status.code", otlpStringValue(statusCode)))
```
Comment (darccio) [inline ddtrace/tracer/stats_to_otlp_metrics.go]: Confirm the `status.code` value format and that it should be ungated across both modes.

`status.code` is emitted as `"Ok"` / `"Error"` for every data point, in both default and OTel-semantics mode. Two things worth confirming:

1. **Value format.** The metric name and bucket bounds in this file are explicitly "matching OTel Span Metrics Connector defaults" (see the `spanMetricBounds` comment). The Span Metrics Connector, however, emits `status.code` using the stringified span-status enum (`STATUS_CODE_UNSET` / `STATUS_CODE_OK` / `STATUS_CODE_ERROR`), not `"Ok"`/`"Error"`. If we're aligning with that connector, this value format is inconsistent with it; if we're following SEMCON-1093 instead, a pointer to the spec would help.

2. **Mode gating.** Every other attribute here is either an OTel-semconv name (emitted in both modes) or a `datadog.*` name (default mode only). `status.code` doesn't obviously belong to either camp — it's a bare key with Datadog-flavored values. Is emitting it verbatim in OTel mode intended?

Not blocking, but since a downstream consumer will group by this dimension, it's worth pinning the exact string values against the spec before this ships.

File: ddtrace/tracer/stats_to_otlp_metrics.go (hunk 4 above)
Comment (darccio) [reply ddtrace/tracer/stats_to_otlp_metrics.go:116]: Also, `SplitN(tag, ":", 2)` on a trailing-colon tag like `entrypoint.name:` yields `["entrypoint.name", ""]` — `len == 2` and `parts[0] != ""`, so we emit `datadog.entrypoint.name = ""`. The guard checks the key is non-empty but not the value, so malformed tags become empty-string attributes. Consider `parts[1] != ""` as well (or decide empty values are acceptable and note it).

File: ddtrace/tracer/stats_to_otlp_metrics.go [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+			pts = append(pts, dp)
+		}
+	}
+	return pts
+}
+
+func decodeAndBuildDataPoint(gs *pb.ClientGroupedStats, sketchBytes []byte, startNs, endNs uint64, isError bool, exactCount uint64, defaultService string, otelMode bool) *otlpmetrics.HistogramDataPoint {
+	bucketCounts, sum, minSec, maxSec, sketchCount, err := sketchToHistogram(sketchBytes, spanMetricBounds[:])
+	if err != nil {
+		log.Error("stats_to_otlp_metrics: failed to decode sketch: %v", err.Error())
```
Comment (darccio) [inline ddtrace/tracer/stats_to_otlp_metrics.go]: Sketch decode failure drops the data point invisibly — reconsider the log level and observability.

On a decode error this logs and returns `nil`, so that OK-or-Error data point silently vanishes from the histogram while the rest of the payload proceeds. Dropping one bad point rather than failing the whole payload is the right call, but two refinements:

- **Log level.** A single malformed/incompatible sketch is arguably `Warn`, not `Error` — at `Error` a steady trickle of decode failures (e.g. a sketch-format skew) would be noisy without being individually actionable.
- **Silent undercount.** The only signal that points were dropped is the log line; the emitted metric just reads low. If this path is reachable in practice, a drop counter (statsd/telemetry) would make the data loss observable rather than requiring someone to notice log volume. This function is currently pure with no statsd handle, so it'd need plumbing — flagging as a design consideration for the wire-up PR rather than a must-fix here.

Minor: `log.Error("...%v", err.Error())` can just pass `err` to `%v`.

**Review by darccio — CHANGES_REQUESTED** [review]

File: ddtrace/tracer/stats_to_otlp_metrics_test.go [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+	hist := m[0].GetHistogram()
+	require.NotNil(t, hist)
+	assert.Equal(t, otlpmetrics.AggregationTemporality_AGGREGATION_TEMPORALITY_DELTA, hist.AggregationTemporality)
+
+	require.Len(t, hist.DataPoints, 1)
+	dp := hist.DataPoints[0]
+	assert.Equal(t, spanMetricBounds[:], dp.ExplicitBounds)
+	assert.Equal(t, len(spanMetricBounds)+1, len(dp.BucketCounts))
+	require.NotNil(t, dp.Sum)
+	assert.Equal(t, uint64(1), dp.Count)
```
Comment (darccio) [inline ddtrace/tracer/stats_to_otlp_metrics_test.go:168]: A few coverage gaps worth closing, most importantly the Count/BucketCounts invariant.

The suite is thorough on attribute mapping and bucket placement. Gaps I noticed:

1. **`Count == Σ BucketCounts`** is never asserted. Given `Count` and `BucketCounts` come from independent sources (see the discussion on `stats_to_otlp_metrics.go`), a test that builds a payload where `Hits`/`Errors` diverge from the sketch's value count — and asserts they either match or are reconciled — would lock in whatever semantics we choose.

2. **`host.name` present branch** is untested — every test passes `ReportHostname=false` (line 245 only covers omission). A case with `ReportHostname=true` + non-empty hostname would cover lines 92-95 of the source.

3. **Multi-bucket payloads** — `makePayload` only ever builds a single `ClientStatsBucket`, so the per-bucket `StartTimeUnixNano`/`TimeUnixNano` threading and the bucket loop are exercised with n=1 only. One payload with two buckets would cover distinct timestamps.

4. **`rpc.response.status_code` string fallback** — the non-numeric `GRPCStatusCode` branch (source lines 186-188) has no test; only the numeric path is covered.

None block the PR, but #1 is the one I'd add given the invariant question.

File: ddtrace/tracer/stats_to_otlp_metrics.go [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+		attrs = append(attrs, otlpKeyValue("http.request.method", otlpStringValue(gs.HTTPMethod)))
+	}
+	if gs.HTTPStatusCode != 0 {
+		attrs = append(attrs, otlpKeyValue("http.response.status_code", otlpIntValue(int64(gs.HTTPStatusCode))))
+	}
+	if gs.HTTPEndpoint != "" {
+		attrs = append(attrs, otlpKeyValue("http.route", otlpStringValue(gs.HTTPEndpoint)))
+	}
+	if gs.GRPCStatusCode != "" {
+		if code, err := strconv.ParseInt(gs.GRPCStatusCode, 10, 64); err == nil {
```
Comment (darccio) [inline ddtrace/tracer/stats_to_otlp_metrics.go:191]: `rpc.response.status_code` can be emitted as either an int or a string for the same attribute key.

When `GRPCStatusCode` parses as an integer it's emitted as an int value; otherwise it falls back to a string value (lines 186-188). Emitting one attribute key with two different value types across data points can trip schema-strict OTLP backends, which may expect a stable type per attribute key.

In practice gRPC status codes are always numeric ("0".."16"), so the string branch is effectively unreachable for well-formed input — which raises the question of whether the fallback is needed at all. Options: drop the string fallback (and log/skip on parse failure, consistent with how other malformed inputs are handled), or always emit as string. Either keeps the type stable. Low severity given real gRPC codes always parse.

**Review by darccio — CHANGES_REQUESTED** [review]

File: ddtrace/tracer/stats_to_otlp_metrics.go (hunk 6 above)
Comment (rachelyangdog) [reply ddtrace/tracer/stats_to_otlp_metrics.go:160]: Got it. I fixed it in a following commit. We now use the sketch-derived count for `dp.Count` as well, so `sum(BucketCounts) == Count` holds by construction. Dropped the `exactCount` parameter from `decodeAndBuildDataPoint` entirely; the exact Hits/Errors integers are no longer involved in the data point. I also added a test that asserts the invariant explicitly.

File: ddtrace/tracer/stats_to_otlp_metrics.go (hunk 8 above)
Comment (rachelyangdog) [reply ddtrace/tracer/stats_to_otlp_metrics.go]: I addressed both in following commits. On value format, I think you're right that "Ok"/"Error" was wrong so I  switched to intValue using the OTel SpanStatus enum integers (`0 = STATUS_CODE_UNSET` for non-error, `2 = STATUS_CODE_ERROR`), which is what the spec header in the system-test file explicitly requires ("Status-code … must be typed OTLP values (intValue / boolValue)"). On mode gating, the ungated is intentional. The  `status.code` is the only signal for identifying error data points in OTel-semantics mode where datadog.* attributes are suppressed.

File: ddtrace/tracer/stats_to_otlp_metrics.go (hunk 9 above)
Comment (rachelyangdog) [reply ddtrace/tracer/stats_to_otlp_metrics.go]: I addressed all three. I changed to `log.Warn` and dropped the redundant .`Error()` call. The drop counter is a good idea and I filed it as a follow-up for the wire-up PR where the statsd handle is available to plumb through.

File: ddtrace/tracer/stats_to_otlp_metrics.go (hunk 7 above)
Comment (rachelyangdog) [reply ddtrace/tracer/stats_to_otlp_metrics.go:176]: In OTel-semantics mode the operation name is absent by design. `gs.Name` is a Datadog concept with no OTel semconv equivalent, so it would only pollute the attribute set. `span.name = Resource` is what the system tests assert (FR07) and also the RFC. I added a comment to the source explaining this for more clarification.

File: ddtrace/tracer/stats_to_otlp_metrics_test.go (hunk 10 above)
Comment (rachelyangdog) [reply ddtrace/tracer/stats_to_otlp_metrics_test.go:168]: I added all four. The Count/BucketCounts test now asserts `sum(BucketCounts) == Count` directly. It also implicitly confirms the fix since both values now come from the sketch. The string fallback test was updated to assert the attribute is absent (we now drop non-numeric codes rather than changing the attribute type). `Host.name` present and multi-bucket timestamp tests cover the previously untested branches.

File: ddtrace/tracer/stats_to_otlp_metrics.go (hunk 11 above)
Comment (rachelyangdog) [reply ddtrace/tracer/stats_to_otlp_metrics.go:191]: Fixed! All paths now emit stringValue. Known codes map to their canonical names ("OK", "NOT_FOUND", etc.), unknown numeric codes emit the numeric string, and non-numeric inputs are silently dropped rather than changing the attribute type. The type is stable across all data points.

**Review by darccio — APPROVED** [review]
LGTM
